core/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java

```diff
@@ -136,9 +154,25 @@ public void onNewInfo(ClusterInfo info) {
             }
             if (reroute) {
                 logger.info("rerouting shards: [{}]", explanation);
-                // Execute an empty reroute, but don't block on the response
-                client.admin().cluster().prepareReroute().execute();
+                reroute();
+            }
+            indicesToMarkReadOnly.removeIf(index ->
+                state.getBlocks().indexBlocked(ClusterBlockLevel.WRITE, index)
+            );
+            if (indicesToMarkReadOnly.isEmpty() == false) {
+                markIndicesReadOnly(indicesToMarkReadOnly);
             }
         }
     }
+
+    protected void markIndicesReadOnly(Set<String> indicesToMarkReadOnly) {
+        // set read-only block but don't block on the response
+        client.admin().indices().prepareUpdateSettings(indicesToMarkReadOnly.toArray(Strings.EMPTY_ARRAY)).
```

Comment:
I think we should protect for errors here and log a warning. It doesn't seem like there are protection higher up the call stacks.